### PR TITLE
FEA Spacewind (real delta P experience)

### PR DESCRIPTION
**Pull request by Kapu1178**
## About the PR 
Implements fairly robust spacewind to Goon's build of FEA.

The way it works is high pressure deltas enqueue movables to the spacewind controller. If they fail `PrepareAirflow()` they will not be queued.

~~The object is then pushed/pulled away/towards the source of the airflow on an interval set in the spacewind controller, currently 0.5 seconds. This repeats until the object's airflow speed drops to 0, at which point the object is removed from the process list.~~

The object is pushed/pulled away/towards the source of the airflow in a sleep loop until it's speed drops to 0 via speed decay, or it is unable to be moved anymore, such as due to magboots being turned on.

## Why's this needed? 
It's fuggin' fun and cool.

## Changelog

```changelog
(u)Kapu1178
(*)Added high pressure airflow ("Spacewind"), that moves and tosses things around!
```

### CODE REVIEW

**Comment by frawhst** [comment]
delta pee when

File: code/modules/atmospherics/FEA_spacewind.dm [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+		return FALSE
+
+	if (!origin || src.airflow_speed < 0 || src.last_airflow_movement > world.time - AIRFLOW_MOVE_DELAY)
+		return FALSE
+
+	if (src.airflow_speed)
+		src.airflow_speed = delta / max(get_dist(src, origin), 1)
+		return FALSE
+
+	///A movement occurs here, but it doesn't mean the mob will be queued to the airflow loop.
```
Comment (ZeWaka) [inline code/modules/atmospherics/FEA_spacewind.dm:131]:
```suggestion
	// A movement occurs here, but it doesn't mean the mob will be queued to the airflow loop.
```

File: code/modules/atmospherics/FEA_spacewind.dm [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+		boutput(src, "<span class='notice'>You are pushed away by a rush of air!</span>")
+
+	src.last_airflow_movement = world.time
+
+	//Get the distance from the flow origin to create some feeling of natural falloff
+	var/airflow_falloff = 9 - get_dist(src, origin)
+	if (airflow_falloff < 1)
+		return FALSE
+
+	///At this point, we're locked into an airflow movement.
```
Comment (ZeWaka) [inline code/modules/atmospherics/FEA_spacewind.dm:145]:
```suggestion
	// At this point, we're locked into an airflow movement.
```

File: code/modules/atmospherics/FEA_spacewind.dm [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+
+/atom/movable/bump(atom/A)
+	if(src.airflow_speed > 0 && src.airflow_origin && src.movement_by_airflow)
+		var/turf/T = get_turf(A)
+		if(airflow_speed > 1)
+			airflow_hit(A)
+			A.airflow_hit_act(src)
+		else if(istype(src, /mob/living/carbon/human/normal) && ismovable(A) && (A:pre_airflow_density == 0))
+			boutput(src, "<span class='notice'>You are pinned against [A] by airflow!</span>")
+			src:changeStatus("stunned", 3 SECONDS) // :)
```
Comment (ZeWaka) [inline code/modules/atmospherics/FEA_spacewind.dm:159]: type it

File: code/modules/atmospherics/FEA_spacewind.dm [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+
+/mob/living/carbon/human/airflow_hit(atom/A)
+	if(istype(A, /obj/structure) || istype(A, /turf/simulated/wall))
+		if(src.airflow_speed > 10)
+			src.changeStatus("stun", (round(src.airflow_speed * 1 SECONDS) + 3))
+			loc.add_blood(src)
+			src.visible_message(
+				"<span class='alert'>[src] splats against \the [A]!</span>",
+				"<span class='subtle'>You hear a loud thud.</span>"
+			)
```
Comment (ZeWaka) [inline code/modules/atmospherics/FEA_spacewind.dm:194]: i don't hear anything.

File: code/modules/atmospherics/FEA_spacewind.dm [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+
+	src.changeStatus("weakened", weak_amt SECONDS)
+
+/obj/airflow_hit_act(atom/movable/flying)
+	. = ..()
+	if(flying.pre_airflow_density == TRUE)
+		src.visible_message(
+			"<span class='alert'>A flying [flying.name] slams into \the [src]!</span>",
+			"<span class='subtle'>You hear a loud slam!</span>"
+		)
```
Comment (ZeWaka) [inline code/modules/atmospherics/FEA_spacewind.dm:239]: ditto

File: code/modules/atmospherics/FEA_spacewind.dm [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+
+///Called when src collides with A during airflow
+/atom/movable/proc/airflow_hit(atom/A)
+	SHOULD_CALL_PARENT(TRUE)
+	airflow_speed = 0
+	airflow_origin = null
+
+///Called when "flying" calls airflow_hit() on src
+/atom/proc/airflow_hit_act(atom/movable/flying)
+	return
```
Comment (ZeWaka) [inline code/modules/atmospherics/FEA_spacewind.dm:184]: you could decrease __health

File: code/modules/atmospherics/FEA_spacewind.dm [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+			break
+
+		LAGCHECK(LAG_MED)
+
+		src.airflow_speed = min(src.airflow_speed, AIRFLOW_SPEED_MIN)
+		src.airflow_speed -= AIRFLOW_SPEED_DECAY
+
+		if(src.airflow_speed > AIRFLOW_SPEED_SKIP_CHECK)
+			if(src.airflow_time++ >= src.airflow_speed - AIRFLOW_SPEED_SKIP_CHECK)
+				sleep(1)
```
Comment (ZeWaka) [inline code/modules/atmospherics/FEA_spacewind.dm:81]: 1 what

**Review by ZeWaka — CHANGES_REQUESTED** [review]
couldn't look at the west stuff

**Comment by Laboredih123** [comment]
how have i not seen this

**Comment by Laboredih123** [comment]
very cool

**Comment by deathrobotpunch** [comment]
could you make it that if someone opens both airlocks at the same time everyone nearby gets sucked out into space?

**Comment by deathrobotpunch** [comment]
it could lead to cool spacing chambers that sec could use for executions

**Comment by TheTyrant25** [comment]
is this still being worked on perchance

**Comment by Laboredih123** [comment]
it was not :(

**Comment by TheTyrant25** [comment]
got permission to work on this, so i may pick it up

**Pull request closed without merging**